File: web/test/parity.test.js
// 파이썬판과 결과가 같은지 대조한다.
//
// 기준 파일 fixtures/parity.txt 는 실제 영상으로 검증을 끝낸 파이썬 구현이
// 만든 정답지다(`python3 tests/make_parity_golden.py` 로 다시 만들 수 있다).
// 두 구현이 갈라지면 이 테스트가 먼저 알려 준다.

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

import { parse, render } from '../src/srt.js';
import { tidy } from '../src/postprocess.js';

const MARKER = '========';

function loadCases() {
  const raw = readFileSync(fileURLToPath(new URL('./fixtures/parity.txt', import.meta.url)), 'utf8');
  const cases = [];
  let current = null;
  let section = '';

  for (const line of raw.split('\n')) {
    if (line.startsWith(`${MARKER} case `)) {
      current = { name: line.slice(`${MARKER} case `.length), note: '', input: '', expected: '' };
      section = '';
    } else if (line.startsWith(`${MARKER} note `)) {
      current.note = line.slice(`${MARKER} note `.length);
    } else if (line === `${MARKER} input`) {
      section = 'input';
    } else if (line === `${MARKER} expected`) {
      section = 'expected';
    } else if (line === `${MARKER} end`) {
      cases.push(current);
      section = '';
    } else if (section) {
      current[section] += `${line}\n`;
    }
  }
  return cases;
}

test('대조할 경우가 충분히 있다', () => {
  assert.ok(loadCases().length >= 10);
});

test('파이썬판과 결과가 같다', () => {
  for (const testCase of loadCases()) {
    const actual = render(tidy(parse(testCase.input)));
    assert.equal(actual, testCase.expected, `[${testCase.name}] ${testCase.note}`);
  }
});

test('대화 표시(-) 띄어쓰기 규칙이 파이썬판과 같다 — 자막의 관습을 따른다', async () => {
  const { tidy } = await import('../src/postprocess.js');

  const run = (texts) =>
    tidy(texts.map((text, i) => ({ startMs: i * 1000, endMs: (i + 1) * 1000 - 1, text })))
      .map((cue) => cue.text);

  // 대부분 띄어쓰는 자막 → 붙은 것도 띄운다.
  const spaced = ['- 그래요', '- 가자', '- 응', '- 네', '-음', '- 좋아', '- 알았어', '- 그럼'];
  assert.equal(run(spaced)[4], '- 음');

  // 모두 붙여 쓰는 자막 → 그대로 둔다 (실측: 어떤 자막은 49줄 모두 붙여 쓴다).
  const tight = ['-안 돼요', '-로지타!', '-네', '-그래'];
  assert.deepEqual(run(tight), tight);

  // 음수는 어느 쪽이든 건드리지 않는다.
  assert.equal(run(['- 그래요', '- 가자', '- 응', '-5도 아래'])[3], '-5도 아래');

  // 대시가 아예 없으면 아무 일도 없다.
  assert.deepEqual(run(['그래요', '가자']), ['그래요', '가자']);
});

File: android/core/src/test/kotlin/com/kgc/subex/core/ParityTest.kt
package com.kgc.subex.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * 파이썬판과 결과가 같은지 대조한다.
 *
 * 기준 파일 `parity.txt` 는 실제 영상으로 검증을 끝낸 파이썬 구현이 만든
 * 정답지다(`python3 tests/make_parity_golden.py` 로 다시 만들 수 있다).
 * 두 구현이 갈라지면 이 테스트가 먼저 알려 준다.
 */
class ParityTest {

    private data class Case(val name: String, val note: String, val input: String, val expected: String)

    private fun loadCases(): List<Case> {
        val raw = requireNotNull(javaClass.getResourceAsStream("/parity.txt")) {
            "parity.txt 가 없습니다. python3 tests/make_parity_golden.py 를 먼저 실행하세요."
        }.bufferedReader(Charsets.UTF_8).readText()

        val cases = mutableListOf<Case>()
        var name = ""
        var note = ""
        val input = StringBuilder()
        val expected = StringBuilder()
        var section = ""

        for (line in raw.split('\n')) {
            when {
                line.startsWith("$MARKER case ") -> {
                    name = line.removePrefix("$MARKER case ")
                    input.setLength(0)
                    expected.setLength(0)
                    section = ""
                }
                line.startsWith("$MARKER note ") -> note = line.removePrefix("$MARKER note ")
                line == "$MARKER input" -> section = "input"
                line == "$MARKER expected" -> section = "expected"
                line == "$MARKER end" -> {
                    cases += Case(name, note, input.toString(), expected.toString())
                    section = ""
                }
                else -> when (section) {
                    "input" -> input.append(line).append('\n')
                    "expected" -> expected.append(line).append('\n')
                }
            }
        }
        return cases
    }

    @Test
    fun goldenFileHasCases() {
        assertTrue(loadCases().size >= 10, "대조할 경우가 너무 적습니다")
    }

    @Test
    fun matchesPythonImplementation() {
        val failures = mutableListOf<String>()
        for (case in loadCases()) {
            val actual = Srt.render(Postprocess.tidy(Srt.parse(case.input)))
            if (actual != case.expected) {
                failures += buildString {
                    append("[").append(case.name).append("] ").append(case.note).append('\n')
                    append("  파이썬: ").append(case.expected.replace("\n", "\\n")).append('\n')
                    append("  코틀린: ").append(actual.replace("\n", "\\n"))
                }
            }
        }
        assertEquals(emptyList(), failures, "파이썬판과 결과가 다릅니다:\n" + failures.joinToString("\n"))
    }

    private companion object {
        const val MARKER = "========"
    }
}
